Render lms main navigation (tabs) with template

StraNNiKK opened the pull request:
The reason for this pull requests is the conversation with @singingwolfboy in Slack:
https://openedx.slack.com/archives/theming/p1449519743000017

Someone wants to rename the “Courseware” tab in the LMS navigation and for now there is no ability to make it easy (only one way to do it is to override the entire template).

With this little improvement every tab could be redefined (for example in microsite theme) with adding specific template (by default courseware/tab_default.html will be taken):
```
courseware/tab_courseware.html 
courseware/tab_info.html 
courseware/tab_discussion.html 
courseware/tab_wiki.html 
courseware/tab_progress.html
courseware/tab_instructor.html
```

singingwolfboy commented:
@StraNNiKK: we use a rebase workflow with the Open edX codebase, not a merge workflow. You have three merge commits in this pull request. Can you take them out? The easiest way to do that would probably be to reset your branch to the tip of `edx/master` and then cherry-pick the one non-merge commit that you have in this pull request, and then force-push to update the pull request.

This page might also be helpful: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

edx-webhook commented:
Thanks for the pull request, @StraNNiKK! I've created [OSPR-1044](https://openedx.atlassian.net/browse/OSPR-1044) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](http://open.edx.org/sites/default/files/wysiwyg/individual-contributor-agreement.pdf)     or indicated your institutional affiliation.     You haven't added yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo. Please create a new pull request to do so!     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

StraNNiKK commented:
@singingwolfboy I've removed merge commits, please check

singingwolfboy commented on common/djangoapps/pipeline_mako/templates/static_content.html:
```diff
@@ -99,3 +99,18 @@
     tmpl.render_context(context)
 %></%def>
+
+<%def name="optional_include_mako_with_fallback(file, fallback, with_microsite=False, **kwargs)"><%
```
Why not just fold this functionality into the `optional_include_mako` function that already exists? If the file isn't found, check and see if a fallback was specified. If so, try to use it; if not, pass.

Also, what happens if the template author wants to define a list of fallbacks?

StraNNiKK replied:
@singingwolfboy do you suggest to extend optional_include_mako function with fallback and kwargs params? It seemed that in slack discussion you first offered to create separated function like optional_include_mako_with_fallback with additional arguments.
Ok I will extend existed function with new arguments.

> Also, what happens if the template author wants to define a list of fallbacks?
Do you suggest to provide fallback param as a list of pathes to templates? I can do it but I think that one fallback template will be sufficient for the 99% cases.

singingwolfboy replied:
You're right that I suggested that in chat, but I was throwing out a suggestion without really considering the implications. Now that I see the code, I think it would be better to fold this improvement into the original function. (Sometimes my off-the-cuff suggestions are wrong. Sorry! :stuck_out_tongue_winking_eye:)

You're right that one fallback template will probably be sufficient in many cases, but probably not *all* cases. We should consider how we want to handle those other cases. I'm open to suggestions -- maybe we should optimize for the "one fallback" case, and have a separate argument that is used for multiple fallbacks. Maybe we shouldn't handle multiple fallbacks at all, and let the template author write the logic into the template. But I suspect there's something we can do to make it at least a *little* bit cleaner.

StraNNiKK replied:
AS for me there are two possible variants: 1. shouldn't handle multiple fallbacks at all (template author should provide provide the logic into his template), 2. to consider fallback argument as a list of templates.
The variant with two separate arguments (one for the first fallback and one for the others) looks strange. Also we could consider fallback param as one fallback if it was passed as s string and as a group of fallbacks if it was passed as a list. But I think this variant strange and non-intuitive as well.

dpowell replied:
FWIW, I like the solution of the fallback argument being either a string or a list of strings (which could of course have a single element.)

StraNNiKK replied:
@dpowell string_or_list argument could be a solution but for me it looks like a little bit unpythonic. The philosophy of python is based on the couple of principles. "Explicit is better than implicit" one of them. So as for my opinion if function always expects an argument as a list it is more explicit (even this list consists of only one element).

StraNNiKK commented:
@singingwolfboy @dpowell I've updated the pull requests according to your comments. About fallback argument - I've left it without changes.

singingwolfboy commented on lms/templates/courseware/course_navigation.html:
```diff
@@ -86,21 +86,11 @@
       % for tab in get_course_tab_list(request, course):
         <%
-            tab_is_active = (tab.tab_id == active_page) or (tab.tab_id == default_tab)
+        tab_is_active = (tab.tab_id == active_page) or (tab.tab_id == default_tab)
```
Nitpick: this can be shortened to:
```python
tab_is_active = tab.tab_id in (active_page, default_tab)
```

StraNNiKK replied:
ok, I'll change this string. But this behavior [was before](https://github.com/edx/edx-platform/blob/master/lms/templates/courseware/course_navigation.html#L89) my pull request :-)

singingwolfboy commented on lms/templates/courseware/tab_default.html:
```diff
@@ -0,0 +1,19 @@
+    <a href="${tab.link_func(course, reverse) | h}" class="${tab_class}">
+    ${_(tab.name) | h}
+    % if tab_is_active:
+        <span class="sr">, current location</span>
```
"current location" needs to be translated

StraNNiKK replied:
I think it shouldn't. Because as the previous one this behavior also [was before](https://github.com/edx/edx-platform/blob/master/lms/templates/courseware/course_navigation.html#L95) the current pull request. And it seems that developers have made it for a some reason. I don't know why but it seems that it is make sense.

mattdrayer replied:
@cptvitamin can you comment on the need for i18n for screen reader span elements?

cptvitamin replied:
@mattdrayer if I understand the question, by not translating the screen reader text, we are assuming that all blind users speak english ;)

singingwolfboy commented on common/djangoapps/pipeline_mako/templates/static_content.html:
```diff
@@ -88,14 +88,19 @@
+        except TemplateLookupException:
+            pass
+if tmpl:
+    tmpl.render_context(context._locals(kwargs))
```
Rather than adding template context using `**kwargs`, can we match the behavior of Mako's [`<%include>`](http://docs.makotemplates.org/en/latest/syntax.html#include) tag as much as possible? `<%include>` uses an `args` parameter, which takes key-value pairs. For example: `<%static:optional_include_mako file="template.html" args="foo='bar'">`

StraNNiKK replied:
@singingwolfboy after the little research and some several attempts I couldn't find the way how to do this. <%include> is a build in function but optional_include_mako isn't. So in case if I use
```
<%static:optional_include_mako file="courseware/tab_${tab.tab_id}.html" fallback="courseware/tab_default.html" with_microsite="True" args="tab=tab,tab_is_active=tab_is_active,tab_class=tab_class" />
```
nothing will be passed into optional_include_mako function.
I've tried to define function optional_include_mako like this:
```
<%def name="optional_include_mako(file, with_microsite=False, fallback=None, args=None, **kwargs)">
```
or like this:
```
<%def name="optional_include_mako(file, with_microsite=False, fallback=None, *args, **kwargs)">
```
but nothing has changed.

jbarciauskas commented:
@nedbat FYI - not sure if this should go to you or someone else, just FYI

dpowell commented:
What are the next steps on this pull request?

jbarciauskas commented:
@dpowell Since DB has left, we just need to figure out who the right person is to pick up this review. It might be @nedbat but he's pretty heavily involved in getting Dogwood out the door at the moment.

nedbat commented:
We should give this to T&L.

cahrens commented:
As I indicated in the JIRA ticket, I think this PR should be reviewed by @mattdrayer on the Solutions team.

mattdrayer commented:
@ziafazal @saleem-latif -- please take a look at this PR when you have a moment

ziafazal commented on lms/templates/courseware/course_navigation.html:
```diff
@@ -86,21 +86,11 @@
     <ol class="course-tabs">
       % for tab in get_course_tab_list(request, course):
```
I think we need to simply things. We should provide ability to override all course tabs instead of each single tab with a separate template file.
we can include a tabs file here like this
`<%static:optional_include_mako file="courseware/tabs.html" >`
Then inside `tabs.html` we should handle rendering of tabs like this
```
 <ol class="course-tabs">
        % for tab in get_course_tab_list(request, course):
...................
...................
...................
</ol>
```

ziafazal commented:
@mattdrayer @StraNNiKK this PR lacks test coverage in specially with course tab overrides.

StraNNiKK commented:
@ziafazal I've already created the test "test_optional_include_mako_with_fallback": https://github.com/edx/edx-platform/pull/11076/files#diff-63196f493f5c1eaf461be34d34bf5320R18
Or do you mean that there will be needed tests in case of overriding course tabs instead of each single tab?

ziafazal commented:
@StraNNiKK yeah tests for tab overrides.

mattdrayer commented:
@StraNNiKK looks like you need a rebase as well

saleem-latif commented on common/djangoapps/pipeline_mako/templates/static_content.html:
```diff
@@ -88,14 +88,19 @@
 </%def>
 
-<%def name="optional_include_mako(file, with_microsite=False)"><%
+<%def name="optional_include_mako(file, with_microsite=False, fallback=None, **kwargs)"><%
```
`with_microsite` was renamed to `is_theming_enabled` in another PR, make sure to update this variable name if you are using it. I think it is one of the conflicts github is showing.

jbarciauskas commented:
Moved to waiting on author, @saleem-latif @StraNNiKK do you need guidance on the variable naming issue?

srpearce commented:
@nedbat @mattdrayer @jbarciauskas @saleem-latif @StraNNiKK Did all of you about the changes that are coming this week? The Courseware tab name is changing to Course and Course Info is changing to Home, and the two tabs are changing places. For more information, see https://open.edx.org/announcements/changes-main-lms-page-coming-soon. 

I don't know if this change affects this PR, but wanted to make sure you knew about it.

StraNNiKK commented:
@mattdrayer @saleem-latif @ziafazal @jbarciauskas @dpowell  I've updated the pull requests: moved all tabs into the one separated template and added tests

ziafazal commented on lms/templates/courseware/course_navigation.html:
```diff
@@ -83,27 +78,11 @@
+     <%
+      tab_list = get_course_tab_list(request, course)
+      %>
+     <%static:optional_include_mako file="courseware/tabs.html" is_theming_enabled="True"
```
@StraNNiKK I'm wondering what could be the reason for using `optional_include_mako` for including course tabs instead of built-in include since course tabs are not optional part of courseware page?.

StraNNiKK replied:
@ziafazal I don't understand what do you mean. I need to override "courseware/tabs.html" in my microsite custom theme. If you have other solution how to do this without using **optional_include_mako** please provide this solution for me

ziafazal replied:
@StraNNiKK what's wrong with including it like this `<%include file="courseware/tabs.html" />` ?

StraNNiKK replied:
@ziafazal it doesn't work with microsite overridden template. The primary goal of the current PR for our project is to have ability to change tabs text/html in the LMS navigation

ziafazal replied:
@StraNNiKK you can try this one to make it work for microsites also
`<%include file="${static.get_template_path('courseware/tabs.html')}" />`

StraNNiKK replied:
@ziafazal seems that it works. thanks! updated the PR again

ziafazal commented on lms/templates/courseware/course_navigation.html:
```diff
@@ -83,27 +78,14 @@
+      base_tabs_tpl = 'courseware/tabs.html'
+      tabs_tmpl = static.get_template_path(base_tabs_tpl)
+      if tabs_tmpl == base_tabs_tpl:
+          tabs_tmpl = '/' + base_tabs_tpl
```
@StraNNiKK what is the reason for prepending `/` to template path. I think it should work without it.

StraNNiKK replied:
@ziafazal I've tried but it doesn't work. The main problem is that **<%include** works only with the slash in the beginning of the line (example: https://github.com/edx/edx-platform/blob/90f2fb868e0646733fad8ec615af2d2dec11801e/lms/templates/static_htmlbook.html#L122 )
But including templates from microsites works through the *optional_include_mako* that works only with relative pathes (example: https://github.com/edx/edx-platform/blob/90f2fb868e0646733fad8ec615af2d2dec11801e/lms/templates/main.html#L125 )
otherwise TemplateLookupException will be raised

ziafazal replied:
@StraNNiKK I have issued PR #11548 to fix absolute path issue with `get_template_path` so you should be able to use it without putting that clutter logic of prepending `/. You should be able to use it like this in both microsite and non microsite cases.
`<%include file="${static.get_template_path('/courses/tabs.html')}" args="tab_list=tab_list,active_page=active_page,default_tab=default_tab,tab_image=tab_image" />
`

ziafazal commented on lms/templates/courseware/course_navigation.html:
```diff
@@ -83,27 +78,14 @@
+          tabs_tmpl = '/' + base_tabs_tpl
+      tab_list = get_course_tab_list(request, course)
+      %>
+     <%include file="${tabs_tmpl}" args="tab_list=tab_list,active_page=active_page,default_tab=default_tab,tab_image=tab_image" />
```
@StraNNiKK I don't think you need to pass everything in `args`. `courseware/tabs.html` should have the same context as the container template.

StraNNiKK replied:
@ziafazal it doesn't work.
Exception **'Undefined' object is not iterable** will always be raised.
As I could see the same approach used in many templates in edx. Example:
https://github.com/edx/edx-platform/blob/90f2fb868e0646733fad8ec615af2d2dec11801e/cms/templates/base.html#L59

StraNNiKK commented:
@mattdrayer @saleem-latif @ziafazal @jbarciauskas What are the next steps with this PR?

nedbat commented:
jenkins run all

ziafazal commented:
jenkins run bokchoy

ziafazal commented:
jenkins run all

StraNNiKK commented:
@ziafazal I've updated the PR according your comments but now the test:
```
./manage.py lms test --verbosity=2 lms/djangoapps/courseware/tests/test_microsites.py:TestMicrosites.test_microsite_course_custom_tabs --settings=test
```
is always failed. So I'm waiting for your fix WL-320. As soon as it will be in master, I'll rebase it with my branch and this test will be set right.

Also I'm fixed the failed DiscussionHomePageTest:
```
paver test_bokchoy -t discussion/test_discussion.py:DiscussionHomePageTest --fasttest
```
but because of this I had to move <ol> </ol> tags from the tabs.html template

StraNNiKK commented:
@ziafazal I've rebased the PR with the last master (with your `get_template_path` fix). Could you please run tests again ?

mattdrayer commented:
jenkins run all

edx-webhook commented:
Hello @StraNNiKK: We are unable to continue with review of your submission at this time. Please see the associated JIRA ticket for more explanation.

mattdrayer commented:
jenkins run all

benpatterson commented:
jenkins ok to test

benpatterson commented:
jenkins add to whitelist

StraNNiKK commented:
@mattdrayer @saleem-latif @ziafazal @jbarciauskas @benpatterson all tests are green. Is there anything else I can do with this PR?

jbarciauskas commented:
@StraNNiKK I'll put this back into awaiting review. @mattdrayer @ziafazal who will give the final thumbs up on this?

mattdrayer commented:
I'll leave the final thumbs up to @ziafazal on this one -- he has the most context at this point.

ziafazal commented:
LGTM :+1:

jbarciauskas commented:
k, who's going to click the green button? :) @ziafazal @mattdrayer
